### ISSUE TITLE
Text editor not working as expected (with USB keyboard)

### ISSUE THREAD

**Issue by f1yn** [koreader/koreader#12709]
* KOReader version: 2024.07
* Device: Kobo Elipsa (1)

#### Issue
I've been hoping to transition to my Elipsa as a typing tool for later in the evenings. 

In the Text Editor plugin, I've noticed:

- I am unable to type ANY special characters with an USB keyboard (ampersands, hash, dollar signs, plus minus, e.c.t). This is expected in sensitive parts of a frontend, but in a text editor I should be allowed to type in those types of characters. In my use case, I'm hoping to edit raw markdown, though not being able to type a dollar sign is wild.

  I've tested on a few USB keyboards at this point, and this issue is affecting all of them.

- Likely related to the above, I also noted that normal text navigation does not work. Control left/right, alt left right, control backspace. As a developer, I've relied on the ergonomics of these shortcuts, and I'm confused why they haven't been added here.

  The up and down navigation keys also don't navigate the text buffer in a normal way. This is very noticeable in monospace than normal text.

- The text cursor is horrendous. I type moderately fast, which means my eyes are facing the screen as a type and not down at my hands. The way the buffer gets chopped up until there's a notable denounce period is not ideal for my use case. I don't think this can be fixed easily outside of disabling the cursor conditionally.

#### Steps to reproduce

Edit a text file with a USB keyboard. Try typing anything that's not alphanumeric. Also try using a whole line delete shortcut.

##### `crash.log` (if applicable)
`crash.log` is a file that is automatically created when KOReader crashes. It can normally be found in the KOReader directory:

* `/mnt/private/koreader` for Cervantes
* `koreader/` directory for Kindle
* `.adds/koreader/` directory for Kobo
* `applications/koreader/` directory for Pocketbook

Android logs are kept in memory. Please go to [Menu] → Help → Bug Report to save these logs to a file.

Please try to include the relevant sections in your issue description.
You can upload the whole `crash.log` file (zipped if necessary) on GitHub by dragging and dropping it onto this textbox.

If your issue doesn't directly concern a Lua crash, we'll quite likely need you to reproduce the issue with *verbose* debug logging enabled before providing the logs to us.
To do so, go to `Top menu → Hamburger menu → Help → Report a bug` and tap `Enable verbose logging`. Restart as requested, then repeat the steps for your issue.

If you instead opt to inline it, please do so behind a spoiler tag:
<details>
  <summary>crash.log</summary>

```
<Paste crash.log content here>
```
</details>

**Comment by f1yn** [comment]
As an aside, how does the Koreader plugin archetecture work?

I mostly everything that I tracts with the UI is Lua, does this mean I could also just edit the plugin over USB and then reload koreader (and apply those changes)?

**Comment by f1yn** [comment]
As an aside, how does the Koreader plugin archetecture work?

I noted everything that interacts with the UI is Lua, does this mean I could also just edit a plugin over usb and then reload koreader (and apply those changes)?

**Comment by Frenzie** [comment]
There currently isn't anything in place to support things like keyboard layouts.

> I noted everything that interacts with the UI is Lua, does this mean I could also just edit a plugin over usb and then reload koreader (and apply those changes)?

Correct, as long as it's nothing but Lua that'll do the trick.

**Comment by NiLuJe** [comment]
> * I've tested on a few USB keyboards at this point, and this issue is affecting all of them.

The ExternalKeyboard plugin enforces a QWERTY US layout, and currently doesn't support extending that.

The original author mentioned that it might be desirable to somehow be able to import xbkd defs, but that never materialized.

Patches welcome ;).

**Comment by NiLuJe** [comment]
> * I also noted that normal text navigation does not work

InputDialog & cie handle re-implement handling of Key events internally, and only very basic navigational stuff is supported.

Patches welcome, but that one might be trickier to get into because the whole thing's sort of messy, and it would need to not break everything else, which takes precedence over a plugin that's only supported on a couple devices ;).

**Comment by f1yn** [comment]
Hey, just reporting back. I've tested _all_ of the the QWERTY keyboards I have over USB (American Layout), and I'm still not able to use `Shift` + any of the numkeys to type symbols.

I felt maybe it was an issue with the plugins, but I noted the same issues within the terminal and other input fields. Turns out, it's the lack of handling of [Shift + alphanumeric keys here](https://github.com/koreader/koreader/blob/fd3727274b44a7ba6531aa778189e0ab3a6e05fa/frontend/ui/widget/inputtext.lua#L709).

I haven't attempted a fix yet, because this is a really low level widget, and mucking about iin here without the right guards (i.e, is this a physical keyboard) - the conditions here might clobber anything else that comes after.

---

As for the cursor, I realized quickly that the logic that handles the cursor drawing is more difficult to debounce as I had hoped (even with the helpful UI debounce helper), so for now: I've commented out the `drawCursor()` calls specifically on my Kobo. I don't suspect I'd be able to debounce the cursor rendering without some heavier refactors to the lua logic that handles that (and even then, that's such a heavy ask considering how many other devices exist).

**Comment by f1yn** [comment]
Is there a non-expenseive way to detect if the event is being emitted from a USB device? Or would I be better of detecting the `Shift` + `1234567890-=` + `Unshift` from the usbkeyboard plugin?

I think the second option might work better, but I'm not fully familiar with how the events are propagating through there.

**Comment by NiLuJe** [comment]
> Is there a non-expensive way to detect if the event is being emitted from a USB device?

No, the input backend doesn't really keep track of which is which.

You might be able to cobble something together to check if *a* USB device is currently connected (by extending `Input:fdopen(fd, path, name)` & `Input:open(fd, path, name)` to carry that over (or, maybe easier, just a dedicated register/unregister/get API), but the actual input code doesn't care which is which.

> Or would it be better off detecting the Shift + 1234567890-= + Unshift from the usbkeyboard plugin?

The only thing the plugin can change is the raw key mapping, actual input and modifier handling is split between the input backend, frontend, the Key class and InputContainer/InputText.

--

TL;DR: No, for your use-case, you'll have to tweak InputText as you found out ;).

**Comment by f1yn** [comment]
I think I'm going to try this from a different angle, but tweaking `InputText` seems to be the lowest level where it can be adjusted. The patch would probably need to be applied where it does the lower-case transference. I'm more inclined to maybe move the conditions around where it will check for different QWERTY keys (not just numeric, all the keys with symbol shifts).

That being said, for my own sanity - I think I might approach things from a different angle. I'm noting that the `inputtext_class` option on the `InputDialog` lets me pass in a custom widget that would be fine for USB keyboards only. Having a plugin tailored explicit for USB keyboard use (and longform content) would probably work better for me needs.

Having a text dialog scroll line by line is not pleasant for longer content, paged text entry would be better for eink - so I can try and give that a go.
